### PR TITLE
Zio-9012 async resumption

### PR DESCRIPTION
**Pull request by eyalfa**
addresses #9012

### CODE REVIEW

**Comment by eyalfa** [comment]
CC: @kyri-petrou

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 1]
```diff
@@ -402,34 +402,40 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
...
+
+            if (supervisor ne Supervisor.none) supervisor.onEnd(exit, self)(Unsafe.unsafe)
+
+            self._runtimeFlags = RuntimeFlags.enable(_runtimeFlags)(RuntimeFlag.WindDown)
 
-          self._runtimeFlags = RuntimeFlags.enable(_runtimeFlags)(RuntimeFlag.WindDown)
+            val interruption = interruptAllChildren()
 
-          val interruption = interruptAllChildren()
+            if (interruption == null) {
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: `eq`?

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 2]
```diff
@@ -402,34 +402,40 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
...
+              effect = interruption.flatMap(_ => exit)(id.location)
             }
-
-            effect = null
-          } else {
-            effect = interruption.flatMap(_ => exit)(id.location)
           }
         } catch {
-          case AsyncJump =>
+          /*case AsyncJump =>
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: leftover?

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 3]
```diff
@@ -402,34 +402,40 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
           val exit =
             runLoop(effect, 0, _stackSize, initialDepth).asInstanceOf[Exit[E, A]]
 
-          if (supervisor ne Supervisor.none) supervisor.onEnd(exit, self)(Unsafe.unsafe)
+          if (null eq exit) {
+            // Terminate this evaluation, async resumption will continue evaluation:
+            effect = null
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:408]: We also need `_forksSinceYield = 0` here

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 4]
```diff
@@ -1188,28 +1221,32 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
...
               cur = null
 
-              while ((cur eq null) && check()) {
+              while (cont && (cur eq null) && check()) {
                 runLoop(iterate.body(), stackIndex, stackIndex, nextDepth) match {
                   case Success(value) =>
                     iterate.process(value)
 
                   case failure =>
+                    cont = failure ne null
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: Let's be explicit here and use `case null => cont = false` instead

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 5]
```diff
@@ -931,16 +937,18 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
     // Note that assigning `cur` as the result of `try` or `if` can cause Scalac to box local variables.
     var cur        = effect
     var done       = null.asInstanceOf[Exit[Any, Any]]
+    var cont       = true
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: Out of curiosity, why did you end up using a flag instead of `return null`? I find that using a return statement is much more explicit on what's the method is going to return, and it's very clear on _when_ it'll return it (i.e., when we we're yielding).

Also, I doubt there is much difference in performance, but any chance we could compare them _just in case_?

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 6]
```diff
@@ -1188,28 +1221,32 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
...
+              }
 
             case yieldNow: ZIO.YieldNow =>
               updateLastTrace(yieldNow.trace)
               if (yieldNow.forceAsync || !stealWork(currentDepth)) {
                 inbox.add(FiberMessage.YieldNow)
                 inbox.add(FiberMessage.resumeUnit)
 
-                throw AsyncJump
+                //throw AsyncJump
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: leftover

**Review by kyri-petrou — COMMENTED** [review]
Looks good overall, just some minor comments below. Although I really think we should use a `return` statement in order to be explicit with returning a null. Otherwise we might be introducing a bug in the future where we end up yielding unintended.

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 5 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: tbh, it never crossed my mind 😎 
however, like many(?) other scala devs I consider the `return` statement 'banned', I recall someone told me it was implemented using an exception and in general (at least at the time) not really recommended.

I guess I can try changing the `YieldNow` case to use return and measure

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 3 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:408]: good catch!

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 4 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: this might have performance implications, can we settle for a verbose comment?

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 5 above)
Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: > I recall someone told me it was implemented using an exception

That's only partially true. This happens only in cases where the `return` keyword is used within a function, e.g., `def foo: () => Int = () => return () => 1`. This is called a "non-local return", and it's indeed implemented via an exception. The Scala 3 compiler raises a warning when this happens, and in Scala 2.13 it can be made into a compiler error by using the `-Xlint:nonlocal-return` compiler flag.

In fact, local returns (the "good" kind) are used extensively in the Scala Std library. E.g., the `List#contains` method is implemented as below, which is perfectly fine.

```scala
  override final def contains[A1 >: A](elem: A1): Boolean = {
    var these: List[A] = this
    while (!these.isEmpty) {
      if (these.head == elem) return true
      these = these.tail
    }
    false
  }
```

In short, the "problem" with `return` which led to it being villainised is that it breaks functional programming semantics (and performance due to non-local returns). However, in imperative-style coding (which is how the `FiberRuntime` is written to improve performance), using `return` is perfectly ok.

**Comment by eyalfa** [comment]
@kyri-petrou ,
attempted the return approach, seems to work well (at least for this benchmark)
so /i took it further and completely eliminated the `done` variable and moved solely to `return`s

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 7]
```diff
@@ -1027,17 +1032,20 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
 
               val result = runLoop(effect.first, stackIndex, stackIndex, currentDepth + 1)
 
-              stackIndex -= 1
-              popStackFrame(stackIndex)
+              if (null eq result)
+                return null
+              else {
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:1037]: The `else` statement is redundant here

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 8]
```diff
@@ -1054,22 +1062,25 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
               stackIndex = pushStackFrame(effect, stackIndex)
 
               val result = runLoop(effect.first, stackIndex, stackIndex, currentDepth + 1)
+              if (null eq result)
+                return null
+              else {
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:1067]: As above

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 9]
```diff
@@ -1123,11 +1136,16 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
                     currentDepth + 1
                   )
 
-                  stackIndex -= 1
-                  popStackFrame(stackIndex)
+                  if (null eq exit)
+                    return null
+                  else {
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala:1141]: As above

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 10]
```diff
@@ -1223,7 +1243,8 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
       }
     }
 
-    done
+    //unreachable
+    ???
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: If we're going to go down this path, consider throwing an `IllegalStateException` with a helpful message to help future devs with debugging (I know it's impossible after a `while(true)` but someone will very likely change the condition in the future)

**Review by kyri-petrou — COMMENTED** [review]
Looks good overall. My previous comments didn't intend for us to use `return` for everything, but AFAICT the code is easier to follow(?). How are benchmarks looking after these changes? I'm quite interested in the ones measuring sync performance as well, you never know how the C2 compiler will like these changes 👀

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 10 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: will do

**Comment by eyalfa** [comment]
@kyri-petrou can u compose a short list of relevant benchmarks?

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 7 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala:1037]: I personally prefer using the explicit `else` in this cases, makes it easier to track the code's execution pass, far easier to figure out u're in a conditional branch than spotting a return statement few lines above

**Comment by kyri-petrou** [comment]
I think just the usual suspects (NarrowFlatmap, BroadFlatmap) should do

**Comment by eyalfa** [comment]
@kyri-petrou 
## baseline:
```info] Benchmark                                             (depth)  (size)   Mode  Cnt       Score       Error  Units
AsyncConcurrentBenchmarks.zioAsyncConcurent               N/A     N/A  thrpt   15    1434.018 ±    19.344  ops/s
BroadFlatMapBenchmark.zioBroadFlatMap                      20     N/A  thrpt   30   18645.377 ±   718.103  ops/s
NarrowFlatMapBenchmark.zioNarrowFlatMap                   N/A    1000  thrpt   30  416595.399 ± 12528.270  ops/s
 AsyncResumptionBenchmark.zioAsyncResumptionBenchmark      N/A     N/A   avgt   15     967.286 ±    71.087  ns/op
```

## rc
```[info] Benchmark                                             (depth)  (size)   Mode  Cnt       Score       Error  Units
AsyncConcurrentBenchmarks.zioAsyncConcurent               N/A     N/A  thrpt   15    1881.474 ±   123.197  ops/s
BroadFlatMapBenchmark.zioBroadFlatMap                      20     N/A  thrpt   30   18681.144 ±   705.810  ops/s
NarrowFlatMapBenchmark.zioNarrowFlatMap                   N/A    1000  thrpt   30  425947.795 ± 23852.585  ops/s
AsyncResumptionBenchmark.zioAsyncResumptionBenchmark      N/A     N/A   avgt   15     276.742 ±    93.871  ns/op
```

**Review by kyri-petrou — APPROVED** [review]
LGTM 👍 Thanks!

I've been playing around with improving scheduling post-yield and got something promising. I'll do a followup PR soon

**Comment by 987Nabil** [comment]
@eyalfa you removed all usage of AsycJump, but not the class
